scripts/split_sentencepiece_ids.awk

```diff
@@ -0,0 +1,61 @@
+# Read (from stdin) the result of encoding a text file with a trained
+# SentencePiece model (as id's).  Write all lines containing  fewer than
+# (the input variable) MAX_LENGTH ids into separate output files, one for
+# each length up to MAX_LENGTH, named FILE_PREFIX0.txt, ..., FILE_PREFIXk.txt
+# (where k is the value of MAX_LENGTH) except that no empty files will be
+# written.  Input lines containing more than MAX_LENGTH ids will be split
+# into non-overlapping chunks with sizes chosen uniformly at random between
+# (the input variables) MIN_CHUNK and MAX_CHUNK (inclusive), with any ids
+# remaining at the end of the line discarded.  Note that the input variables
+# should satisfy the constraints 1 <= MIN_CHUNK <= MAX_CHUNK <= MAX_LENGTH.
+#
+# Example invocation:
+#   cat input.txt | awk -f split_sentencepiece_ids.awk \
+#     -v MAX_LENGTH=300 -v MIN_CHUNK=15 -v MAX_CHUNK=30 \
+#     -v FILE_PREFIX=data/sentencepiece/en.spm_encoded_ids_
+
+# Return a random integer in [m, n].
+function randint(m, n) {
+    return m + int((n + 1 - m) * rand())
+}
+
+BEGIN {
+    if (MIN_CHUNK < 1) {
+        print ("Error: MIN_CHUNK (" MIN_CHUNK ") < 1.") > "/dev/stderr"
+        exit 1
+    }
+    if (MAX_CHUNK < MIN_CHUNK) {
+        print ("Error: MAX_CHUNK (" MAX_CHUNK ") < MIN_CHUNK (" MIN_CHUNK ").") > "/dev/stderr"
+        exit 1
+    }
+    if (MAX_LENGTH < MAX_CHUNK) {
+        print ("Error: MAX_LENGTH (" MAX_LENGTH ") < MAX_CHUNK (" MAX_CHUNK ").") > "/dev/stderr"
+        exit 1
+    }
+}
+
+{
+    if (NF <= MAX_LENGTH) {
+        filename = FILE_PREFIX NF ".txt"
+        print $0 > filename
+    } else {  # break long inputs into shorter chunks
+        n_ids = split($0, ids)
+        i_start = 1  # awk indexing is 1-based
+        while (i_start <= n_ids) {
+            i_end = i_start + randint(MIN_CHUNK, MAX_CHUNK) - 1
+            if (i_end > n_ids) {
+                i_end = n_ids
+            }
+            len = i_end - i_start + 1
+            if (len < MIN_CHUNK) {  # don't write any final, runt, chunk
```

Comment:
Is this just discarding the tokens from the end, instead of choosing the boundaries within constraints that satisfy the desired chunk lengths?

Reply:
To provide an example of what I'm hoping for: a sentence that is 75 tokens long could be divided into chunks of 35 and 40, or 39 and 36, or any pair of lengths within the constraints, but not 70 and 5.

Reply:
To answer your first question, yes, the code you reviewed might discard tokens at the ends of lines.

Do I have this straight:  You don't care so much about the distribution of lengths of chunks, but you do want to avoid discarding any tokens when splitting long sentences into chunks.  

In general it won't be possible to guarantee that no tokens are ever discarded (e.g. if the minimum chunk size is 35 tokens, the maximum chunk size (and maximum permitted sentence length) is 68, and you get a sentence of 69 tokens, there is no way to satisfy the constraints without dropping some tokens:  you can't use two chunks because that requires 70 or more tokens, and one chunk uses at most 68).  But if the maximum chunk size is at least twice the minimum chunk size minus one, then the following strategy is guaranteed to use all the tokens:  First emit one fewer than the maximum possible number of chunks of the minimum size.  Then the number of tokens remaining will be greater than or equal to the minimum chunk size and strictly less than twice the minimum chunk size.  So if the maximum size is at least twice the minumum size minus one, the remaining tokens will form a chunk of at most the maximum size, and can all be output in one final chunk.

I've implemented code for that strategy and am testing it now.  Please let me know if it seems acceptable to  you.

Reply:
Yes, that's why I said the maximum size should be twice the minimum size. And that effect on the distribution of lengths is fine.